### ISSUE TITLE
Model converges on CPU but not GPU, what could it be?

### ISSUE THREAD

**Issue by KastanDay** [mosaicml/composer#1862]
I'm on a very standard system: x86 EPYC CPU on A100 GPU.

```bash
** Environment **
---------------------------------
System Environment Report
Created: 2023-01-06 11:42:29 CST
---------------------------------

PyTorch information
-------------------
PyTorch version: 1.13.0
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: NVIDIA DGX Server (x86_64)
GCC version: (GCC) 4.8.5 20150623 (Red Hat 4.8.5-44)
Clang version: Could not collect
CMake version: version 2.8.12.2
Libc version: glibc-2.17

Python version: 3.8.13 (default, Oct 21 2022, 23:50:54)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-3.10.0-1160.11.1.el7.x86_64-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: 11.0.221
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: A100-SXM4-40GB
Nvidia driver version: 450.156.00
cuDNN version: Probably one of the following:
/usr/lib64/libcudnn.so.7.6.3
/usr/lib64/libcudnn.so.8.2.1
/usr/lib64/libcudnn_adv_infer.so.8.2.1
/usr/lib64/libcudnn_adv_train.so.8.2.1
/usr/lib64/libcudnn_cnn_infer.so.8.2.1
/usr/lib64/libcudnn_cnn_train.so.8.2.1
/usr/lib64/libcudnn_ops_infer.so.8.2.1
/usr/lib64/libcudnn_ops_train.so.8.2.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] lovely-numpy==0.2.8
[pip3] numpy==1.22.4
[pip3] pytorch-ranger==0.1.1
[pip3] torch==1.13.0
[pip3] torch-optimizer==0.3.0
[pip3] torchaudio==0.13.0
[pip3] torchmetrics==0.9.3
[pip3] torchvision==0.14.0
[conda] blas                      1.0                         mkl
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] lovely-numpy              0.2.8                    pypi_0    pypi
[conda] mkl                       2021.4.0           h06a4308_640
[conda] mkl-service               2.4.0            py38h7f8727e_0
[conda] mkl_fft                   1.3.1            py38hd3c417c_0
[conda] mkl_random                1.2.2            py38h51133e4_0
[conda] numpy                     1.22.4                   pypi_0    pypi
[conda] pytorch                   1.13.0          py3.8_cuda11.6_cudnn8.3.2_0    pytorch
[conda] pytorch-cuda              11.6                 h867d48c_0    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] pytorch-ranger            0.1.1                    pypi_0    pypi
[conda] torch-optimizer           0.3.0                    pypi_0    pypi
[conda] torchaudio                0.13.0               py38_cu116    pytorch
[conda] torchmetrics              0.9.3                    pypi_0    pypi
[conda] torchvision               0.14.0               py38_cu116    pytorch


Composer information
--------------------
Composer version: 0.12.0
Composer commit hash: None
Host processor model name: AMD EPYC 7742 64-Core Processor
Host processor core count: 128
Number of nodes: 1
Accelerator model name: A100-SXM4-40GB
Accelerators per node: 1
CUDA Device Count: 1
```
CUDA: Driver Version: 450.156.00   CUDA Version: 11.0 

* Note via Conda, my `pytorch-cuda` is `11.6` but my system is using CUDA `11.0`. This shouldn't be a big problem, but YMMV.

** To reproduce

1. Instantiate Trainer, and call `trainer.fit()` with below model. 

## Expected behavior

Model converges when `device=cpu` but loss is all `NANs` when device=gpu. 

After several days of thinking and debugging I can't figure out what it could be! 

## Additional context

Note, all model parameters are `torch.float32`.

I'm using a Huggingface model, but it's sufficiently custom that _I believe_ I need to write my own ComposerModel class, instead of using your conversion method `from composer.models import HuggingFaceModel`.

My full model class:
```python
class VPT_model(ComposerModel):
  '''
  Custom VPT implementaiton for MosaicML.
  https://docs.mosaicml.com/en/v0.11.1/trainer/using_the_trainer.html
  '''

  def __init__(self,
               model_huggingface_name: str = "google/t5-v1_1-large"):
    super().__init__()

    self.device = "cuda" if torch.cuda.is_available() else "cpu"
    self.model = T5ForConditionalGeneration.from_pretrained(model_huggingface_name).to(self.device)
    self.model.train()

  def forward(self, batch):
    input_embeds_arr = torch.cat(
        [batch['clip_pooled_embedding'], batch['clip_last_hidden_states'], batch['caption_embedding']],
        dim=1)  # concat along sequence dimension
    return self.model.forward(inputs_embeds=input_embeds_arr, attention_mask=batch['attn_mask_arr'], labels=batch['labels'])

  def loss(self, outputs, batch):
    return outputs[0].sum() # same results with and without .sum()

  def eval_forward(self, batch, outputs=None):
    '''
    Docs: https://docs.mosaicml.com/en/v0.11.1/api_reference/generated/composer.ComposerModel.html#composer.ComposerModel.eval_forward
    todo: Placeholder so that training doesn't crash.
    '''
    return 0
```

Thank you for any help you can provide!

**Comment by eracah** [comment]
Hi Kastan, 
VPT on youtube? Sounds like a cool project.

May I see your call to `Trainer`? 
I suspect changing the precision to 'fp32' (default for cpu, but not for gpu) may be the fix: `Trainer(..., precision='fp32',)`

Also, one thing to note: it looks like your system CUDA version is older than the CUDA version used to build pytorch. I think it's usually recommended that the system CUDA is the same or newer than the pytorch CUDA version.

Let me know if that helps!

**Comment by KastanDay** [comment]
Problem solved!! Phew, thank you!

Yes, very excited about the project. It won 'best project of the year' in UIUC's advanced NLP seminar: https://github.com/KastanDay/video-pretrained-transformer
